lib/src/server.ml

```diff
@@ -223,35 +231,110 @@ let run config host port root trim_period trim_size =
+          let f writer = Async.return @@ Async.Writer.write writer contents in
+          Blocks.put ~f t (hash h) executable
         in
         Async_rpc.implement Rpc.block_put f
       and index_get =
-        let f t (path, h) =
+        let f { t; _ } (path, h) =
+          let* () = Logs_async.info (fun m -> m "GET index/%s" h) in
           Blocks.get ~path ~f:Core.In_channel.read_lines t (hash h) >>| function
           | Some (v, _) -> Some v
           | None -> None
         in
         Async_rpc.implement Rpc.index_get f
       and index_put =
-        let f t (path, h, contents) =
-          Blocks.put ~f:Core.Out_channel.output_lines ~path t (hash h) false
-            contents
+        let f { t; _ } (path, h, contents) =
+          let* () = Logs_async.info (fun m -> m "PUT index/%s" h) in
+          let f writer =
+            Async.return
+            @@ List.iter ~f:(Async.Writer.write_line writer) contents
+          in
+          Blocks.put ~f ~path t (hash h) false
         in
         Async_rpc.implement Rpc.index_put f
+      and metadata_put =
+        let f { t; connection } (h, payload) =
+          let* () = Logs_async.info (fun m -> m "PUT metadata/%s" h) in
+          let h = hash h in
+          match Cache.Local.Metadata_file.of_string payload with
+          | Result.Ok { contents; _ } ->
+            let+ () =
+              if Config.ranges_include ranges h then
+                let f writer =
+                  Async.return @@ Async.Writer.write writer payload
+                in
+                Blocks.put ~f t h false
+              else
+                Async.Deferred.return ()
+            and+ () =
+              match contents with
+              | Files files ->
+                let f { Cache.File.digest; _ } =
+                  if Config.ranges_include ranges digest then
+                    let h = Digest.to_string digest in
+                    let* () = Logs_async.info (fun m -> m "FETCH %s" h) in
+                    Async.Rpc.Pipe_rpc.dispatch Rpc.block_get connection h
+                    >>= Rpc.decode_block_get
+                    >>= function
+                    | Some (contents, executable) ->
+                      let f writer =
+                        Async.Pipe.transfer ~f:Core.Fn.id contents
+                          (Async.Writer.pipe writer)
+                      in
+                      Blocks.put ~f t (hash h) executable
+                    | None -> Async.return ()
+                  else
+                    Async.Deferred.return ()
+                in
+                Async.Deferred.List.iter ~f files
+              | Value _ -> Async.return ()
+            in
+            ()
+          | Result.Error e -> failwith e
+        in
+        Async_rpc.implement Rpc.metadata_put f
       in
       match
         Async.Rpc.Implementations.create
           ~implementations:
-            [ block_get; block_has; block_put; index_get; index_put ]
+            [ block_get
```

Comment:
It seems worth factoring out definitions of all these implementations into separate functions and document what they are doing. I think I could guess what they do most of the time, but writing down short specs would make it much easier for me to review the code, since I'd then know the intended behaviour too (not only the actual one).